src/frontend/hooks/useLocation.ts

```diff
@@ -1,112 +1,136 @@
-  minDistanceInterval?: number;
-  /** Minimum time to wait between each update in milliseconds. */
-  minTimeInterval?: number;
-  /** Maximum time to wait between each update in milliseconds - updates could take longer than this, but if a new location is available we will always get it after this time */
-  maxTimeInterval?: number;
-}
+  LocationState,
+  useLocationStore,
+} from '../contexts/LocationStoreContext';
 
-export interface LocationState {
-  location: LocationObject | undefined;
-  error: Error | undefined;
+type Accuracy = 'better' | 'worse' | 'unchanged' | 'newBounds';
+
+/**
+ * Represents the changes in relevant location information since the last time the _hook subscriber_ accepted a new location update
+ */
+export interface ParameterChanges {
+  /**
+   * Distance change in meters
+   */
+  distance: number;
+  /**
+   * Amount of time elapsed in milliseconds
+   */
+  time: number;
+  /**
+   * Whether the accuracy has improved, become worse, or remained the same
+   */
+  accuracy: Accuracy;
 }
 
-export function useLocation({
-  minDistanceInterval: distanceInterval = 1,
-  minTimeInterval,
-  maxTimeInterval,
-  maxDistanceInterval,
-}: LocationOptions): LocationState {
-  const [location, setLocation] = React.useState<LocationState>({
-    location: undefined,
-    error: undefined,
-  });
-
-  const [permissions] = useForegroundPermissions();
-
-  useEffect(() => {
-    if (!permissions || !permissions.granted) return;
-
-    let ignore = false;
-    const locationSubscriptionProm = watchPositionAsync(
-      {
-        accuracy: Accuracy.BestForNavigation,
-        distanceInterval,
-      },
-      debounceLocation({
-        minTimeInterval,
-        maxTimeInterval,
-        maxDistanceInterval,
-      })(location => {
-        if (ignore) return;
-        setLocation({location, error: undefined});
-      }),
+type ShouldAcceptUpdateCheck = (changes: ParameterChanges) => boolean;
+
+const DEFAULT_SHOULD_ACCEPT_UPDATE: ShouldAcceptUpdateCheck = () => true;
+
+/**
+ * Hook for subscribing to location updates and optionally providing a callback to control the frequency of stateful updates received by the hook subscriber
+ *
+ * @example
+ * ```
+ * // Outside your component...
+ * function shouldAcceptLocationUpdate(changes) {
+ *   // Accept location update if hook subscriber if it has been at least 5 seconds or 10 meters away since it accepted the last one
+ *   return changes.time >= 5_000 || changes.distance >= 10 || accuracy === 'better'
+ * }
+ *
+ *
+ * // Inside your component...
+ * const locationState = useLocation(shouldAcceptLocationUpdate);
+ * ```
+ */
+export function useLocation(
+  shouldAcceptUpdate: ShouldAcceptUpdateCheck = DEFAULT_SHOULD_ACCEPT_UPDATE,
+) {
+  const locationStore = useLocationStore();
+  const prevState = useRef<LocationState | undefined>(
+    locationStore.getSnapshot(),
+  );
+
+  const getSnapshot = useCallback(() => {
```

Comment:
Logic here looks sound. Note that you could have also kept the existing debounce to avoid the arguably more confusing `shouldAcceptUpdate` function e.g.

```
const getSnapshot = useCallback(() => {
  return debounceLocation(options)(location => location)
}, Object.values(options))
```

Although would need new logic for minDistanceInterval, and could change the signature so you can just `return debounceLocation(options)`.